core/dbt/task/deps.py

```diff
@@ -59,40 +121,166 @@ def track_package_install(
+
+        if packages_yml:
+            with open(packages_yml_filepath, "w") as pkg_obj:
+                pkg_obj.write(
+                    yaml.dump(packages_yml, Dumper=dbtPackageDumper, default_flow_style=False)
+                )
+
+                fire_event(
+                    DepsAddPackage(
+                        package_name=self.args.package,
+                        version=self.args.version,
+                        packages_filepath=packages_yml_filepath,
+                    )
+                )
+
+    def lock(self) -> None:
+        lock_filepath = f"{self.project.project_root}/{PACKAGE_LOCK_FILE_NAME}"
+
         packages = self.project.packages.packages
+        packages_installed: Dict[str, Any] = {"packages": []}
+
         if not packages:
             fire_event(DepsNoPackagesFound())
             return
 
         with downloads_directory():
-            final_deps = resolve_packages(packages, self.project, self.cli_vars)
+            resolved_deps = resolve_packages(packages, self.project, self.cli_vars)
+
+        # this loop is to create the package-lock.yml in the same format as original packages.yml
+        # package-lock.yml includes both the stated packages in packages.yml along with dependent packages
+        for package in resolved_deps:
+            lock_entry = _create_packages_yml_entry(
+                package.name, package.get_version(), package.source_type()
+            )
+            packages_installed["packages"].append(lock_entry)
+        packages_installed[PACKAGE_LOCK_HASH_KEY] = _create_sha1_hash(
+            self.project.packages.packages
+        )
+
+        with open(lock_filepath, "w") as lock_obj:
+            yaml.safe_dump(packages_installed, lock_obj)
+
+        fire_event(DepsLockUpdating(lock_filepath=lock_filepath))
+
+    def run(self) -> None:
+        if self.args.ADD:
+            self.add()
+        else:
+            if self.args.dry_run:
+                raise dbt.exceptions.DbtRuntimeError(
+                    "Invalid flag `--dry-run` when not using `--add`."
+                )
+
+        # Check lock file exist and generated by the same pacakges.yml
+        # or dependencies.yml.
+        lock_file_path = f"{self.project.project_root}/{PACKAGE_LOCK_FILE_NAME}"
+        if not system.path_exists(lock_file_path):
+            self.lock()
+        elif self.args.upgrade:
+            self.lock()
+        else:
+            # Check dependency definition is modified or not.
+            current_hash = _create_sha1_hash(self.project.packages.packages)
+            previous_hash = load_yml_dict(lock_file_path).get(PACKAGE_LOCK_HASH_KEY, None)
+            if previous_hash != current_hash:
+                self.lock()
 
+        # Early return when dry run or lock only.
+        if self.args.dry_run or self.args.lock:
+            return
+
+        if system.path_exists(self.project.packages_install_path):
```

Comment:
This is a change, update

Reply:
Yes! IMO this is a big improvement (sorry I never saw/responded to this sooner)